### PR TITLE
Feature: Unified Domain Interface

### PR DESCRIPTION
**Pull request by nmboavida**

### CODE REVIEW

File: sources/standards/composability/items.move [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    ) {
+        let authority_type = type_name::get<Auth>();
+        let idx = get_or_insert_authority_idx(authority_type, items);
+
+        let nft_id = object::id(&child_nft);
+        vec_map::insert(&mut items.nfts, nft_id, idx);
+        dof::add(&mut items.id, nft_id, child_nft);
+    }
+
+    // TODO: Reassess when domain model is restructured
```
Comment (bausano) [inline sources/standards/composability/items.move:167]: TODO

File: sources/standards/composability/items.move [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    /// Borrows `Items` from `T`
+    ///
+    /// #### Panics
+    ///
+    /// Panics if `T` was not composed within the `Items`.
+    public fun borrow_items(parent_uid: &UID): &Items {
+        df::borrow(parent_uid, utils::marker<Items>())
+    }
+
+    // TODO: This is unsafe because there is no access control...
```
Comment (bausano) [inline sources/standards/composability/items.move:88]: TODO

File: sources/standards/display.move [hunk 3]
```diff
@@ -1,238 +1,97 @@
...
-    use nft_protocol::witness;
-    use nft_protocol::nft::{Self, Nft};
     use nft_protocol::witness::Witness as DelegatedWitness;
     use nft_protocol::collection::{Self, Collection};
 
     /// Witness used to authenticate witness protected endpoints
     struct Witness has drop {}
 
-    struct DisplayDomain has drop, store {
+    struct DisplayInfo has drop, store {
```
Comment (bausano) [inline sources/standards/display.move:17]: Do we still want to have this domain once display standard is applied to our protocol?

File: sources/standards/domain_bag.move [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+    /// Domain already defined
+    ///
+    /// Call `collection::borrow` to borrow domain
+    const EEXISTING_DOMAIN: u64 = 2;
+
+    /// Transaction sender not logical owner
+    const EINVALID_SENDER: u64 = 3;
+
+    // TODO: Consider wrapping DomainBag in option in order to use df instead of dof
```
Comment (bausano) [inline sources/standards/domain_bag.move:30]: Is this TODO relevant to this module?

File: sources/standards/domain_bag.move [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        ctx: &mut TxContext,
+    ) {
+        let domain = new_(ctx);
+
+        df::add(nft_uid, DomainKey {}, domain);
+    }
+
+    // To be used in the context of programmable transactions.
+    // TODO: The problem now is that all domains will have the same access permission,
+    // what if creators want to determine access atomically for each domain?
```
Comment (bausano) [inline sources/standards/domain_bag.move:84]: They can still have per domain permission by using a protected struct as a key with which the dynamic object is stored. That is, have `struct MyDomainKey {} has store` which can only be constructed from the domains module. Then when you access the domain bag's `&mut UID`, other modules cannot query your type unless they have obtained somehow `MyDomainKey` instance.

Comment (bausano) [reply sources/standards/domain_bag.move:84]: Ah, now I remember we're using `Marker` for key storage which can be constructed by anyone.

File: sources/standards/domain_bag.move [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+            domain,
+        )
+    }
+
+    /// Adds domain to `Nft`
+    ///
+    /// #### Panics
+    ///
+    /// Panics if domain already exists.
+    public fun add_domain_delegated<Domain: store>(
```
Comment (bausano) [inline sources/standards/domain_bag.move:219]: Why is this the only API which accepts delegated witness type.

File: sources/utils/utils.move [hunk 7]
```diff
@@ -23,6 +28,32 @@ module nft_protocol::utils {
         10_000
     }
 
+    public fun proof_of_type<T: key + store>(uid: &UID, object: &T): UidType<T> {
```
Comment (bausano) [inline sources/utils/utils.move:31]: `T: key` is sufficient

File: sources/utils/utils.move [hunk 8]
```diff
@@ -15,6 +16,10 @@ module nft_protocol::utils {
     /// Used to mark type fields in dynamic fields
     struct Marker<phantom T> has copy, drop, store {}
 
+    struct UidType<phantom T> has drop {
```
Comment (bausano) [inline sources/utils/utils.move:19]: minor: 
```
public fun uid_type_to_inner(uid: &UidType): ID {
  uid.id
} 
```

File: sources/standards/composability/items.move (hunk 1 above)
Comment (nmboavida) [reply sources/standards/composability/items.move]: Resolved

File: sources/standards/display.move (hunk 3 above)
Comment (nmboavida) [reply sources/standards/display.move]: You're right, I removed it now

File: sources/standards/domain_bag.move [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+    /// fun init(witness: SUIMARINES, ctx: &mut TxContext) {
+    ///     let nft = nft::new(&Witness {}, tx_context::sender(ctx), ctx);
+    /// }
+    /// ```
+    public fun new(ctx: &mut TxContext): DomainBag {
+        new_(ctx)
+    }
+
+    public fun add_domain_bag(
+        nft_uid: &mut UID,
```
[the remark this answers is not in the document]
Comment (nmboavida) [reply sources/standards/domain_bag.move:71]: It's still to be defined I would say but perhaps we keep for any object that wants to neatly organise their OriginByte related fields. Yeah that would include the `Collection`

File: sources/utils/utils.move (hunk 7 above)
Comment (nmboavida) [reply sources/utils/utils.move:46]: Done ;)

File: sources/standards/composability/items.move (hunk 2 above)
Comment (nmboavida) [reply sources/standards/composability/items.move]: Resolved

File: sources/standards/domain_bag.move (hunk 6 above)
Comment (nmboavida) [reply sources/standards/domain_bag.move]: Resolving as it's outdated
